### ISSUE TITLE
Improve how docstrings for strategies appear/are generated?

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#92]
It must be possible to automate the generation of a sub header etc for each strategy?

Currently in the `strategies.rst` file I have manually listed all of them. I can't seem to be able to find how to automate this...

(Here's the currently rendered documentation: http://axelrod.readthedocs.org/en/latest/)

All files are written using sphinx in rst and can be found in the `docs` directory.

**Comment by langner** [comment]
What do you mean by sub header?

**Comment by drvinceknight** [comment]
Sorry, I didn't really explain this well. 

So at the moment: http://axelrod.readthedocs.org/en/latest/strategies.html shows every strategy's docstring listed out.

That is how an excerpt of how this is done in: https://raw.githubusercontent.com/drvinceknight/Axelrod/master/docs/strategies.rst

```
Here are some of the basic strategies:
--------------------------------------

.. currentmodule:: axelrod.strategies
.. autoclass:: Alternator
.. autoclass:: Cooperator
.. autoclass:: Defector
.. autoclass:: Random
.. autoclass:: TitForTat

A list of all further (honest) strategies:
------------------------------------------

.. autoclass:: AntiTitForTat
.. autoclass:: Appeaser
```

So I'm listing each strategy manually, I think I must be missing some sphinx but can't seem to find it: I'd like to just be able to say 'Ok list all the class from `axelrod.strategies`' here. 

*Also* if possible if they could each have their own subheader (`<h4>Alternator</h4>` or something like that) that is slightly nicer than what currently appears (`class axelrod.strategies.Alternator`) that would be cool although perhaps that's not needed... Just trying to think about readability for none coders but maybe that's not important and this isn't exactly 'not readable'. As I've written this I think that that is perhaps really not needed...

**Comment by langner** [comment]
You can use `automodule`, but then you can't easily break the list into the section you have.

**Comment by drvinceknight** [comment]
I tried `automodule` but without success. Let me try again...

**Comment by langner** [comment]
You could have a Python script generate the `rst` file, then you can do whatever you want :D

**Comment by drvinceknight** [comment]
Yeah I think `automodule` picks up doscrtrings for given submodule but not for classes perhaps? Unless I'm doing something wrong: e6d71cc

> You could have a Python script generate the rst file, then you can do whatever you want :D

Had thought that... Any idea how that would play with read the docs? I like that it just automatically runs the sphinx makefile... Perhaps I can just tweak that file... I need to look in to it.

Do you think it's worth looking in to or perhaps nicer to just manually list the files so that we 'have control' so to speak? Easy enough to do for each PR... (Thinking aloud here)

**Comment by drvinceknight** [comment]
This could all just sit in the new `manage.py` script I keep on talking about...

**Comment by langner** [comment]
> Had thought that... Any idea how that would play with read the docs?

As long as the result is the RST file before sphinx starts it's work, it will be independent. See #95.

**Comment by drvinceknight** [comment]
Just seen #95... On my list of 'understand awesome voodoo stuff that @langner has written' that's towards the bottom as makefiles still scare the hell out of me...

**Comment by drvinceknight** [comment]
As far as I can tell, the make script doesn't run on read the docs. This f4f57be14c4d019496920fae3a27014fb5f6b807 and previous added 5 new strategies that don't seem to appear.

Just re-opening the issue in case you had any ideas but I'm happy to close it straight down and simple run `make html` locally :)

**Comment by langner** [comment]
You might need to force those targets. Otherwise, just run `make` (no need for `make html` locally, since you're not interested in getting the actual HTML into the repo).

**Comment by langner** [comment]
Forcing targets in makefiles: https://www.gnu.org/software/make/manual/html_node/Force-Targets.html

**Comment by drvinceknight** [comment]
Ok: thanks, will take a look! :+1:

**Comment by drvinceknight** [comment]
As far as I can tell: running `make` and even `make html` does not update the rst file as required...

At line 26 of the MAKEFILE you have:

```
strategies.rst: strategies.py
        python $< > $@
```

Just checking that that `>` is indeed throwing the output of `strategies.py` to the `rst` file (sorry but I have tried on many occasion to get my head around makefiles and I don't think I've succeeded yet...).

**Comment by drvinceknight** [comment]
(I expect all of the above is me being an idiot and not understanding the voodoo so please do say so :D)

**Comment by drvinceknight** [comment]
The more I think about this the more I think we need to manually curate documentation for each and every strategy, this could actually turn out to be a very nice reference... Would require a docs/strategies/<strategy>.rst file for each one but I think it's worth doing... I'm happy enough to start working my way through it...

**Comment by langner** [comment]
I tend to think that keeping documentation as close to the code as possible is the best, and it can't be closer than a docstring.

**Comment by drvinceknight** [comment]
Yeah I agree with that...

I think I'm just unsatisfied with how this all works so I thought that just manually curating the docs for strategies could be a nice way to go...

(I'm just thinking aloud here...)

It is silly though to have the actual docs anywhere else than in the docstring...

Let me work on a branch on this and get back to you... The docs need a bunch of tlc I think, it would be nice to be able to easily find a description of each strategy instead of just that one big list... (still thinking aloud)

**Comment by langner** [comment]
> Yeah I agree with that...
> 
> I think I'm just unsatisfied with how this all works so I thought that just manually curating the docs for strategies could be a nice way to go...
> 
> (I'm just thinking aloud here...)
> 
> It is silly though to have the actual docs anywhere else than in the docstring...
> 
> Let me work on a branch on this and get back to you... The docs need a bunch of tlc I think, it would be nice to be able to easily find a description of each strategy instead of just that one big list... (still thinking aloud)

I don't understand what final effect you're going for here. Do you simply want each strategy to appear in the table of contents?

Of course, you can also keep additional docs in separate files, if you like. But it makes sense to keep strategy-specific content along with the class.

**Comment by drvinceknight** [comment]
> I don't understand what final effect you're going for here. Do you simply want each strategy to appear in the table of contents?

I'm not sure I understand either... Probably just mumbling. I think after the merge you just did (#147) we're probably all good. :)
